qadence/ml_tools/config.py

```diff
@@ -28,6 +29,80 @@
+        every (int, optional): Callback to be called each `every` epoch. Defaults to 1.
+            If callback_condition is None, we set
+            callback_condition to returns True when iteration % every == 0.
+        call_before_opt (bool, optional): If true, callback is applied before training.
+            Defaults to False.
+        call_end_epoch (bool, optional): If true, callback is applied during training,
+            after an epoch is performed. Defaults to True.
+        call_after_opt (bool, optional): If true, callback is applied after training.
+            Defaults to False.
+        call_during_eval (bool, optional): If true, callback is applied during evaluation.
+            Defaults to False.
+    """
+
+    def __init__(
+        self,
+        callback: Callable[..., None],
+        callback_condition: Callable[..., bool] | None = None,
+        every: int = 1,
+        call_before_opt: bool = False,
+        call_end_epoch: bool = True,
+        call_after_opt: bool = False,
+        call_during_eval: bool = False,
+    ) -> None:
+        """Initialized Callback.
+
+        Args:
+            callback (Callable[..., None]): Callback function accepting an
+                OptimizeResult as ifrst argument.
+            callback_condition (Callable[..., bool] | None, optional): Function that conditions the
+                call to callback. Defaults to None.
+            every (int, optional): Callback to be called each `every` epoch. Defaults to 1.
+                If callback_condition is None, we set
+                callback_condition to returns True when iteration % every == 0.
+            call_before_opt (bool, optional): If true, callback is applied before training.
+                Defaults to False.
+            call_end_epoch (bool, optional): If true, callback is applied during training,
+                after an epoch is performed. Defaults to True.
+            call_after_opt (bool, optional): If true, callback is applied after training.
+                Defaults to False.
+            call_during_eval (bool, optional): If true, callback is applied during evaluation.
+                Defaults to False.
+        """
+        self.callback = callback
+        self.call_before_opt = call_before_opt
+        self.call_end_epoch = call_end_epoch
+        self.call_after_opt = call_after_opt
+        self.call_during_eval = call_during_eval
+
+        self.every = every
+
+        if callback_condition is None:
+            if self.every > 1:
+                self.callback_condition = lambda opt_result: opt_result.iteration % self.every == 0
+            else:
+                self.callback_condition = lambda opt_result: True
+        else:
+            self.callback_condition = callback_condition
+
+    def __call__(self, opt_result: OptimizeResult, *args: Any, **kwargs: Any) -> Any:
+        if self.callback_condition(opt_result):
+            return self.callback(opt_result, *args, **kwargs)
+
```

Comment:
As a general comment here, I would delegate the responsibility of triggering the callback condition to the callback condition itself. Here is a possible option
```python
@dataclass
class TrainingStep:
    """Information about which phase of the training we are in."""
    epoch_number: int | None = None
    before_opt: bool = True
    after_opt: bool = False
    during_eval: bool = False

CallbackFunction = Callable[[OptimizeResults], None]
CallbackCondition = Callable[[TrainingStep], bool]


class Callback:
    def __init__(self, callback_function: CallbackFunction, callback_condition: CallbackCondition):
        self._callback_function = callback_function
        self._callback_condition = callback_condition

    def __call__(self, training_step: TrainingStep, opt_result: OptimizeResult):
        if self.callback_condition(training_step):
            return self._callback_function(opt_result)
```

`TrainingStep` would be an internal 'clock' for the training phase, updated in `train_*` functions. As mentioned, we let `callback_condition` decide on whether the clock is right to trigger. In this way we could also remove `iteration` from `OptmizeResults`.

Reply:
Hm, you can also trigger `callback_condition` based on the loss value or a condition coming from model or optimizer, so that is why I kept also the `callback_condition` to take as input the `OptimizeResult` instance. Although, I don't demonstrate it here directly. Also, the callback would be missing when it is called which would help us filter them before getting in the loop, and avoid unnecessary calling `callback_condition`.